### PR TITLE
chore: fixing context propagation on mongo callback

### LINKED ISSUE #390: mongodb instrumentation: nested span is not generated when using callback
### What version of OpenTelemetry are you using?
    "@opentelemetry/api": "0.18.0"
    "@opentelemetry/core": "0.18.0"
    "@opentelemetry/instrumentation-mongodb": "0.14.0"

### What version of Node are you using?
12.20.0

### What did you do?

```
import { SimpleSpanProcessor } from '@opentelemetry/tracing';
import { NodeTracerProvider } from '@opentelemetry/node';
import { ZipkinExporter } from '@opentelemetry/exporter-zipkin';

import { registerInstrumentations } from '@opentelemetry/instrumentation';
import { HttpInstrumentation } from '@opentelemetry/instrumentation-http';
import { MongoDBInstrumentation } from '@opentelemetry/instrumentation-mongodb';

const serviceName = 'otel-mongodb-test';

const provider = new NodeTracerProvider();

registerInstrumentations({
    instrumentations: [
        new HttpInstrumentation(),
        new MongoDBInstrumentation(),
    ],
    tracerProvider: provider,
});

const zipkinExporter = new ZipkinExporter({
    serviceName,
    url: 'http://localhost:9411/api/v2/spans'
});

provider.addSpanProcessor(new SimpleSpanProcessor(zipkinExporter));
provider.register();
const tracer = provider.getTracer(serviceName);

import * as http from 'http';
import * as mongodb from 'mongodb';

let collection;
const mongodbConnect = async () => {
    try {
        const client = await mongodb.MongoClient.connect('mongodb://localhost:27017');
        collection = client.db('test-db').collection('users');
        console.log('\nSuccessfully connected to mongoDB!\n');
    } catch (err) {
        console.log(`Error while connecting to mongoDB ${err}`, err);
        throw err;
    }
};

mongodbConnect();

http.createServer(function (_, res) {
    const doc = { name: 'Rick' };

    collection.insertOne(doc, function () {
        collection.findOne({}, function () {
            res.end();
        });
    });
}).listen(9090);
```

### What did you expect to see?

The new `mongodb.find` span being generated as a child of the `mongodb.insert` span.

### What did you see instead?
Screenshot 2021-03-17

### Additional context
It happens when a callback for a parent `mongodb.insert` operation is used.

When I'm using `Promise.then` instead of a callback a span for a `find` operation is generated as expected.

The following example works fine:
```
collection.insertOne(doc).then(function () {
        collection.findOne({}, function () {
            res.end();
        });
    });
```

### PR DESCRIPTION
**Pull request by obecny**
## Which problem is this PR solving?

- Fixes #390 - fix is only in instrumentation package as plugin is going to be removed soon anyway

## Short description of the changes

- when using callback instead of promise mongodb is using tick when calling callback. This causing the context to be not propagated correctly

### CODE REVIEW

File: plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts [hunk 1]
```diff
@@ -406,7 +409,10 @@ export class MongoDBInstrumentation extends InstrumentationBase<
         });
       }
       span.end();
-      return resultHandler.apply(this, args);
+
+      return context.with(activeContext, () => {
```
[the remark this answers is not in the document]
Comment (dyladan) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:413]: we could also use `context.with(activeContext, resultHandler, this, ...args)`

[the remark this answers is not in the document]
Comment (obecny) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:413]: Hmm all 3 options would work, so I would like to stay with the gate #1 if you don't mind ;)

**Review by dyladan — APPROVED** [review]

File: examples/mongodb/package.json [hunk 2]
```diff
@@ -33,8 +33,8 @@
     "@opentelemetry/exporter-zipkin": "^0.18.0",
     "@opentelemetry/instrumentation": "^0.18.0",
     "@opentelemetry/instrumentation-http": "^0.18.0",
+    "@opentelemetry/instrumentation-mongodb": "^0.14.0",
```
Comment (jtmalinowski) [inline examples/mongodb/package.json:36]: Could be `^0.15.0`, potentially will fix itself when merging `main`

File: plugins/node/opentelemetry-instrumentation-mongodb/test/mongodb.test.ts [hunk 3]
```diff
@@ -14,8 +14,7 @@
  * limitations under the License.
  */
 
-// for testing locally use this command to run docker
-// docker run -e MONGODB_DB=opentelemetry-tests -e MONGODB_PORT=27017 -e MONGODB_HOST=localhost -p 27017:27017 --name otmongo mongo
+// for testing locally "npm run docker:start" / "npm run docker:stop"
```
Comment (jtmalinowski) [inline plugins/node/opentelemetry-instrumentation-mongodb/test/mongodb.test.ts:17]: It doesn't start detached, so `docker:stop` doesn't make sense.

File: plugins/node/opentelemetry-instrumentation-mongodb/package.json [hunk 4]
```diff
@@ -6,6 +6,8 @@
   "types": "build/src/index.d.ts",
   "repository": "open-telemetry/opentelemetry-js-contrib",
   "scripts": {
+    "docker:start": "docker run -e MONGODB_DB=opentelemetry-tests -e MONGODB_PORT=27017 -e MONGODB_HOST=localhost -p 27017:27017 --name otmongo mongo",
```
Comment (jtmalinowski) [inline plugins/node/opentelemetry-instrumentation-mongodb/package.json:9]: User experience would be better if we got rid of the name (`--name otmongo`) and instead auto-remove (`--rm`). If you run the command as it is, then stop the container, then run again, you get: 
```
docker: Error response from daemon: Conflict. The container name "/otmongo" is already in use by container
```

Comment (obecny) [reply plugins/node/opentelemetry-instrumentation-mongodb/package.json]: thx for tip it works better now

**Comment by vmarchaud** [comment]
cc @jtmalinowski i believe your comments were addressed, are you okay to merge it ?

**Review by jtmalinowski — APPROVED** [review]

**Pull request merged**